src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
Jason, I think I could stop by the office at around 11am on Friday to continue discussing in person.
I’m ready to meet in a middle ground as I mentioned in this PR since day 1, and still go with constructor injection.

Nevertheless, the more I think about it, the more it makes sense to support property injection with the exact same proven and extensible code here: [DefaultControllerPropertyActivator](https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultControllerPropertyActivator.cs)

When we built the unity container, we had lots of this religious wars that took weeks (since the container was our  deliverable after all). Back then I was on the side of taking a stand and not even supporting property and method injection, as most of times it was miss-used and led to user objects that required 2-step initialization (aka temporal coupling) before they are usable. This is not a limitation for us anyway, since we already have a 2-step initialization an users should not consider grains usable until OnActivateAsync is called.

Using property injection by copying the code from MVC means that we don’t put ANY constraint on how users decide to construct objects. It is also as intuitive, but at the same time we are standing on the decisions of the MVC team and their documented semantics (which we are not shoehorning, they are good IMO). It also is a proven approach that works and we can continue evolving (such as allowing to replace the IGrainPropertyActivator if we so decide to support).

Just trying to say to not dismiss a good and already well-thought approach just because “we think constructor injection is cooler than property injection”. We wanted to avoid those religious wars by standing on their decisions, but we started one just because we deviated and want to decide how grains should be resolved.

Just my 2 cents to think a little bit before we can continue the discussion